protocols/mdns/src/behaviour.rs

```diff
@@ -149,138 +277,102 @@ impl NetworkBehaviour for Mdns {
+        while let Poll::Ready(event) = Pin::new(&mut self.if_watch).poll(cx) {
+            let multicast = From::from([224, 0, 0, 251]);
+            let socket = self.recv_socket.get_ref();
+            match event {
+                Ok(IfEvent::Up(inet)) => {
+                    if inet.addr().is_loopback() {
+                        continue;
                     }
-
-                    if !expired.is_empty() {
-                        let event = MdnsEvent::Expired(ExpiredAddrsIter {
-                            inner: expired.into_iter(),
-                        });
-
-                        return Poll::Ready(NetworkBehaviourAction::GenerateEvent(event));
+                    if let IpAddr::V4(addr) = inet.addr() {
+                        log::trace!("joining multicast on iface {}", addr);
+                        if let Err(err) = socket.join_multicast_v4(&multicast, &addr) {
+                            log::error!("join multicast failed: {}", err);
+                        } else {
+                            self.send_buffer.push_back(build_query());
+                        }
+                    }
+                }
+                Ok(IfEvent::Down(inet)) => {
+                    if inet.addr().is_loopback() {
+                        continue;
+                    }
+                    if let IpAddr::V4(addr) = inet.addr() {
+                        log::trace!("leaving multicast on iface {}", addr);
+                        if let Err(err) = socket.leave_multicast_v4(&multicast, &addr) {
+                            log::error!("leave multicast failed: {}", err);
+                        } else {
+                            self.send_buffer.clear();
+                        }
                     }
-                },
-                Poll::Pending => (),
+                }
+                Err(err) => log::error!("if watch returned an error: {}", err),
             }
         }
-
-        // Polling the mDNS service, and obtain the list of nodes discovered this round.
-        let discovered = loop {
-            let service = mem::replace(&mut self.service, MdnsBusyWrapper::Poisoned);
-
-            let packet = match service {
-                MdnsBusyWrapper::Free(service) => {
-                    self.service = MdnsBusyWrapper::Busy(Box::pin(service.next()));
-                    continue;
-                },
-                MdnsBusyWrapper::Busy(mut fut) => {
-                    match fut.as_mut().poll(cx) {
-                        Poll::Ready((service, packet)) => {
-                            self.service = MdnsBusyWrapper::Free(service);
-                            packet
-                        },
-                        Poll::Pending => {
-                            self.service = MdnsBusyWrapper::Busy(fut);
-                            return Poll::Pending;
-                        }
+        // Poll receive socket.
+        while self.recv_socket.poll_readable(cx).is_ready() {
```

Comment:
> whether we can be informed about "read readiness" despite there not yet being a full datagram to read

I'm pretty sure that isn't possible. A linux datagram socket does not become readable until a datagram arrives and not when an ip packet arrives. I can try finding a reference.

> may result in missing wakeups (since now_or_never() does not poll with the cx context)

this is actually not possible, if now_or_never returns None, `poll_readable` is called again. the reason why it is done this way is because most `async` futures cancel and deregister the waker if they are dropped. `poll_readable` was added for exactly that reason, it prevents not getting wakeups.

> but it turns out the datagram we then try to send is too large for the available space in the socket send buffer

it's a datagram socket not a tcp socket. it is guaranteed that a 65k byte datagram (derived from a max udp payload len of u16 + the 8byte header) can be sent. I can also try to find a reference.  

TODO: find some references

Reply:
> this is actually not possible, if now_or_never returns None, poll_readable is called again.

I see. I overlooked that the code is calling `poll_readable` within `while` :facepalm:. That makes it irrelevant whether there may even be false positives. Thanks for the clarification.

> it's a datagram socket not a tcp socket. it is guaranteed that a 65k byte datagram (derived from a max udp payload len of u16 + the 8byte header) can be sent. I can also try to find a reference.

I would be surprised if being notified of write readiness on a datagram socket would be so pessimistic, i.e. only signaling write readiness when a maximally-sized datagram can be sent. My understanding is that write readiness can be signaled whenever there is space in the socket send buffer, however much that may be, but I'm happy to be proven wrong.

Reply:
I do recall that this was a contentious point on the LKML many years ago, but I stopped reading that list around 2012. I suspect that we must assume that the 64kiB send guarantee is not actually upheld by all systems (just like not all systems actually provide a monotonic clock even when they say so … ).